src/selection.coffee

```diff
@@ -487,47 +487,48 @@ class Selection extends Model
   #
   # If there selection spans more than one line, all the lines are joined together.
   joinLines: ->
-    selectedRange = @getBufferRange()
-    if selectedRange.isEmpty()
-      return if selectedRange.start.row is @editor.buffer.getLastRow()
-    else
-      joinMarker = @editor.markBufferRange(selectedRange, invalidationStrategy: 'never')
-
-    rowCount = Math.max(1, selectedRange.getRowCount() - 1)
-    for row in [0...rowCount]
-      @cursor.setBufferPosition([selectedRange.start.row])
-      @cursor.moveToEndOfLine()
-
-      # Remove trailing whitespace from the current line
-      scanRange = @cursor.getCurrentLineBufferRange()
-      trailingWhitespaceRange = null
-      @editor.scanInBufferRange /[ \t]+$/, scanRange, ({range}) ->
-        trailingWhitespaceRange = range
-      if trailingWhitespaceRange?
-        @setBufferRange(trailingWhitespaceRange)
+    @cursor.autoscroll {}, =>
```

Comment:
Do you think we could perform the line join as a single change operation on the buffer? You would have to keep more local state in the method but it would reduce the change events dramatically. Batching everything in the model would be a more universal solution, but it's not a simple change. I'll add some more thoughts about that subject to the PR in a bit. In the meantime, seems like combining here could be a win.